File: baekjoon/recursion/1914-hanoi.cpp
#include <iostream>
#include <string>
#include <cstdio>
#include <cmath>

int cnt = 0;

using namespace std;

void hanoi(int num, int src, int target) {
    if (num == 1) {        
        printf("%d %d\n", src, target);
        return;
    }
    int other = 6 - src - target;
    hanoi(num - 1, src, other);
    printf("%d %d\n", src, target);
    hanoi(num - 1, other, target);
}


int main() {
    int n;
    bool print = false;
    cin >> n;

    //get 2^n - 1 (bigint)
    string s = to_string(pow(2, n));
    int idx = s.find('.');
    s = s.substr(0, idx);
    s[s.length() - 1] -= 1;
    
    cout << s << endl;
    
    if (n <= 20) {
        hanoi(n, 1, 3);
    }    
    return 0;
}

File: baekjoon/recursion/1914-hanoi.py
n = int(input())

def hanoi(source, target, size):
    if size == 1:
        print("{} {}".format(source, target))
        return
    
    other = 6 - source - target
    hanoi(source, other, size - 1)
    print("{} {}".format(source, target))
    hanoi(other, target, size - 1)
    
print(2 ** n - 1)
if n <= 20:
    hanoi(1, 3, n)